File: vigridr/src/client/GameClient.cpp
#include <chrono>
#include <exception>
#include <iostream>
#include <stdio.h>
#include <string>
#include <sys/time.h>
#include <thread>
#include <unistd.h>

#include <gflags/gflags.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>

#include "../thrifts/gen-cpp/Game.h"
#include "ClientLogic.h"

DEFINE_int32(port, 9090, "Port used to connect with the server.");

const char* const kVersion = "v1.1";
const char* const kUsageMessage =
  "This program is the client for mjollnir matches";

using ::apache::thrift::protocol::TBinaryProtocol;
using ::apache::thrift::protocol::TProtocol;
using ::apache::thrift::TException;
using ::apache::thrift::transport::TBufferedTransport;
using ::apache::thrift::transport::TSocket;
using ::apache::thrift::transport::TTransport;

using ::mjollnir::vigridr::Command;
using ::mjollnir::vigridr::GameClient;
using ::mjollnir::vigridr::GameInfo;
using ::mjollnir::vigridr::GameInit;
using ::mjollnir::vigridr::GameStatus;
using ::mjollnir::vigridr::WorldModel;

void playGame(GameClient& client) {
  GameInit gameInit;
  client.ready(gameInit);
  init(gameInit);
  GameInfo gameInfo = gameInit.gameInfo;
  while (true) {
    client.getGameInfo(gameInfo);
    if (gameInfo.gameStatus == GameStatus::FINISHED) {
      endOfGame(gameInfo.gameResult);
      break;
    }
    if (gameInfo.isMyTurn) {
      Command command = playTurn(gameInfo.worldModel, gameInfo.cycle);
      client.sendCommand(command);
    }
  }
}

int main(int argc, char** argv) {
  gflags::SetVersionString(kVersion);
  gflags::SetUsageMessage(kUsageMessage);
  gflags::ParseCommandLineFlags(&argc, &argv, true);

  try {
    boost::shared_ptr<TTransport> socket(new TSocket("localhost", FLAGS_port));
    boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));
    boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
    GameClient client(protocol);
    transport->open();
    playGame(client);
    transport->close();
  } catch (TException &tx) {
    printf("ERROR: %s\n", tx.what());
  }
}

File: vigridr/src/client/GameClient.java
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TTransportException;
import org.apache.thrift.TException;

public class GameClient
{
    public static void playGame(Game.Client client) throws TException {
        GameInit gameInit = client.ready();
        ClientLogic solution = new ClientLogic(gameInit);
        
        GameInfo gameInfo;
        while (true) {
            gameInfo = client.getGameInfo();
            if (GameStatus.FINISHED.equals(gameInfo.getGameStatus())) {
                solution.endOfGame(gameInfo.gameResult);
                break;
            }
            if (gameInfo.isIsMyTurn()) {
                Command command = solution.playTurn(gameInfo.worldModel, gameInfo.moveList, gameInfo.cycle);
                client.sendCommand(command);
            }
        }
    }

    public static void main(String[] args) {
        int port = 9090;
        for (int i = 0; i < args.length; i++) {
            switch(args[i]) {
                case "--port":
                    port = new Integer(args[++i]);
                    break;
                case "--help":
                    System.out.println("--port: Port used to connect with the server. (int)");
                    return;
            }
        }

        try {
            TSocket tSocket = new TSocket("localhost", port);
            TTransport transport = tSocket;
            TProtocol protocol = new TBinaryProtocol(transport);
            Game.Client client = new Game.Client(protocol);
            transport.open();
            playGame(client);
            transport.close();
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}
